jax/interpreters/xla.py

```diff
@@ -490,22 +508,75 @@ def _xla_callable(fun, device, backend, name, *arg_specs):
       extend_name_stack(wrap_name(name, 'jit')), *xla_args)
   built = c.Build(c.Tuple(*out_nodes))
 
+  if device:
+    device_assignment = ((device.id,),)
+  else:
+    device_assignment = xb.get_backend(backend).get_default_device_assignment(nreps, nparts)
+    device_assignment = onp.vectorize(lambda d: d.id)(onp.array(device_assignment))
+
   options = xb.get_compile_options(
-      num_replicas=nreps,
-      num_partitions=1,
-      device_assignment=(device.id,) if device else None)
+      num_replicas=nreps, num_partitions=nparts, device_assignment=device_assignment)
   compiled = built.Compile(compile_options=options, backend=xb.get_backend(backend))
 
-  if nreps == 1:
-    return partial(_execute_compiled, compiled, backend, result_handlers, tuple_args)
-  else:
-    return partial(_execute_replicated, compiled, backend, result_handlers, tuple_args)
+  return create_execute_fn(
+      compiled, device_assignment=device_assignment, backend=backend,
+      tuple_args=tuple_args, result_handlers=result_handlers,
+      out_partition_ids=out_partition_ids)
+
+def create_execute_fn(
+    compiled, *, device_assignment, backend, tuple_args, result_handlers, out_partition_ids):
+  # TODO(cjfj): Expose device assignment from the executable.
+  local_devices = compiled.local_devices()
+  # We are assuming here that `local_devices` are ordered by replica then partition.
+  # TODO(cjfj): Expose logical devices from the executable, rather than re-computing here.
+  local_logical_devices = []
+  local_device_ids = {d.id for d in local_devices}
+  for replica_id, replica_devices in enumerate(device_assignment):
+    for partition_id, device_id in enumerate(replica_devices):
+      if device_id in local_device_ids:
+        local_logical_devices.append((replica_id, partition_id))
+
+  # We take our outputs from the replica 0 devices. Pre-compute the mappings
+  # to these devices for each partition.
+  local_replica0_devices = {
+      part_id: local_device_id for local_device_id, (replica_id, part_id)
+      in enumerate(local_logical_devices) if replica_id == 0}
+
+  def execute(*args):
```

Comment:
I'm a little worried about adding overheads to the non-partition path... what do you think about keeping the old _execute_compiled and _executed_replicated functions for now, with a TODO to benchmark if there's a noticeable difference?

Reply:
Done. Do we have any good benchmarks to try on this?